ffi/tests/ctests/src/tests.c

```diff
@@ -0,0 +1,221 @@
+#ifdef NDEBUG
+#undef NDEBUG
+#endif
+
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+#include <assert.h>
+
+#include <wirefilter.h>
+
+#define WIREFILTER_STRING(v, s) {(v).data = (const unsigned char *)(s); (v).length = strlen(s);}
+
+wirefilter_externally_allocated_str_t wirefilter_init_string(const char *s) {
+	wirefilter_externally_allocated_str_t str = {(const unsigned char *)s, strlen(s)};
+	return str;
+}
+
+void test_wirefilter_ffi_01() {
+	wirefilter_scheme_t *scheme = wirefilter_create_scheme();
+	assert(scheme != NULL);
+	wirefilter_free_scheme(scheme);
+}
+
+void test_wirefilter_ffi_02() {
+	wirefilter_scheme_t *scheme = wirefilter_create_scheme();
+	assert(scheme != NULL);
```

Comment:
I know @RReverser has suggested to use `assert`, however the reason why I was advocating for the proper C testing framework is because currently failed assert will just terminate the whole testing process with the error code and it won't run any consequent tests. 
That might be OK for now, but as API and, thus, test code will grow current approach will be more and more inconvenient  for debugging and root cause investigation.
As the first step it would be nice to have at least proper test function names and assert messages.

Reply:
> As the first step it would be nice to have at least proper test function names and assert messages.

Standard `assert` doesn't really allow you to customise error messages unfortunately, but I suppose line number + proper function name should be fine.

I don't have strong opinion regarding using or not using picotest though.

Reply:
Oh no. I have a horrible idea that actually works.

Rust side:

```rust
#[no_mangle]
unsafe extern fn assert(check: bool, msg: *const std::os::raw::c_char) {
    assert!(check, "{}", std::ffi::CStr::from_ptr(msg).to_str().unwrap());
}

...

#[test]
fn some_test() {
    extern {
        fn some_test();
    }

    unsafe {
        some_test();
    }
}
```

C side:

```c
#include <stdbool.h>

extern void assert(bool check, const char *msg);

...

void some_test() {
	assert(false, "msg");
}
```

Then each test runs in own thread and we're basically reusing Rust framework for managing them fully. Example of error message for above:

```
running 1 test
test some_test ... FAILED

failures:

---- some_test stdout ----
thread 'some_test' panicked at 'msg', src\lib.rs:3:5
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.


failures:
    some_test

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
```

Reply:
I think this is quite nice, because you get full integration with Rust framework which also runs all tests in parallel (like it does for Rust tests) and you get proper stacktraces, messages and so on...

Reply:
I'll try that

Reply:
Its quite nice indeed :+1: 
I also tried to generate the tests using the `paste` crate with something like
```Rust
ffi_ctest!(create_scheme)
```
But as it involves macros in extern block, we was able do so in stable because of https://github.com/rust-lang/rust/issues/49476

Reply:
You don't need to use macro in stable, you can instead generate `extern` block itself using a macro.

Reply:
And you probably don't need an external crate for that, this should work as a simple inline macro.

Reply:
Yep thank you for helping me on that macro :+1: